engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+	if obj.State != StateRunning {
+		return checkOk, nil
+	}
+	// Changes in cpu, location and/or image require a server rebuild.
+	if c, err := obj.checkApplyServerRebuild(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+	} else if !c {
+		checkOk = false
+	}
+	// Changes in rescue mode can be made while the server is running.
+	if c, err := obj.checkApplyRescueMode(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyRescueMode failed")
+	} else if !c {
+		checkOk = false
+	}
+	// TODO: checkapply SSH keys
+	return checkOk, nil
+}
+
+// checkApplyServerState powers the server on/off, creates a new instance, or
+// deletes the existing one as needed, to ensure the correct state.
+func (obj *HetznerRes) checkApplyServerState(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerState(apply: %t)", apply)
+	}
+	var checkOk = true
+	// Make sure the server exists as intended before further checks.
+	// TODO: add create permission
+	if obj.server == nil {
+		if obj.State == StateAbsent {
+			return true, nil
+		}
+		if !apply {
+			return false, nil
+		}
+		checkOk = false
+		obj.init.Logf("requesting server creation")
+		if err := obj.createServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "createServer failed")
+		}
+	}
+	// Make sure the server is in a steady state before further changes.
+	// TODO: test if hetzner api accepts queries while in transient state
+	if err := obj.waitForSteadyState(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "waitForSteadyState failed")
+	}
+	if obj.stateConverged() {
+		return checkOk, nil
+	}
+	// Request changes if the server is not in the intended state yet.
+	// TODO: add permissions (poweron, poweroff, delete)
+	checkOk = false
+	if !apply {
+		return false, nil
+	}
+	switch obj.State {
+	case StateRunning:
+		obj.init.Logf("requesting server poweron")
+		if _, _, err := obj.client.Server.Poweron(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.poweron failed")
+		}
+	case StateOff:
+		obj.init.Logf("requesting server poweroff")
+		if _, _, err := obj.client.Server.Poweroff(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.poweroff failed")
+		}
+	case StateAbsent:
+		obj.init.Logf("requesting server deletion")
+		if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "client.server.delete failed")
+		}
+	}
+	// Wait until the requested changes are confirmed. Error otherwise.
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "getServerUpdate failed")
+	}
+	if err := obj.waitForConvergedState(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "waitForConvergedState failed")
+	}
+	return checkOk, nil
+}
+
+// checkApplyServerRebuild rebuilds the server instance as needed,
```

Comment:
Can you expand in the docstrings somewhere what rebuild means. Eg: is it destroying the user data in that machine? If so, perhaps the user DOES NOT want that and so this should be guarded? For example, sometimes you can't transform a machine into a different type without deleting all the data... The virt resource does this, but *if* it needs to make a destructive change, then you also have to set a flag in the params. In that case "destructive" count as restarting. So data-at-rest loss is even worse. Restarting without expecting or being okay with it is also very bad. See:

```
	// RestartOnDiverge is the restart policy, and can be: `ignore`,
	// `ifneeded` or `error`.
	RestartOnDiverge string `lang:"restartondiverge" yaml:"restartondiverge"`
```

So we give the user the option to ignore, do it if needed, or error the resource because we weren't expecting this. Make sense?

Reply:
Agreed. I like the three options you describe here, but I'm not sure if/how to extend that to other important permissions like creating and deleting servers. Do you consider that to be part of the RestartOnDiverge behavior? In that case I would rename it to something like "AllowDestructiveChanges". Thoughts? 

Creating new servers isn't really destructive, but it can be expensive if abused... Best to add a second permission field? This one seems less urgent though, it's harder to make big mistakes with server creation than deletion.

I also have TODOs for permissions related to powering on/of and enabling/disabling rescue mode. These are all non-destructive changes though, so probably fine to leave as is? Wanted to check now that we're on the topic. 

The most flexible solution I can see is adding a specific private permission for each different operation, and then cluster them at Init() so that you can adjust the coverage of public permissions in one place. If you feel like one or two permission params are enough then it would probably be easier to just hardcode those more plainly.